src/osp/Active/SysHierarchy.cpp

```diff
@@ -0,0 +1,153 @@
+    ActiveEnt child = rReg.create();
+    rReg.emplace<ACompHierarchy>(child);
+
+    if (!name.empty())
+    {
+        rScene.reg_emplace<ACompName>(child, name);
+    }
+
+    set_parent_child(rScene, parent, child);
+
+    return child;
+}
+
+void SysHierarchy::set_parent_child(ActiveScene& rScene, ActiveEnt parent, ActiveEnt child)
+{
+    auto &rReg = rScene.get_registry();
+
+    ACompHierarchy& childHierarchy = rReg.get<ACompHierarchy>(child);
+    ACompHierarchy& parentHierarchy = rReg.get<ACompHierarchy>(parent);
+
+    // if child has an existing parent, cut first
+    if (rReg.valid(childHierarchy.m_parent))
+    {
+        cut(rScene, child);
+    }
+
+    // set new child's parent
+    childHierarchy.m_parent = parent;
+    childHierarchy.m_level = parentHierarchy.m_level + 1;
+
+    // If has siblings (not first child)
+    if (parentHierarchy.m_childCount)
+    {
+        ActiveEnt sibling = parentHierarchy.m_childFirst;
+        auto& siblingHierarchy = rReg.get<ACompHierarchy>(sibling);
+
+        // Set new child and former first child as siblings
+        siblingHierarchy.m_siblingPrev = child;
+        childHierarchy.m_siblingNext = sibling;
+    }
+
+    // Set parent's first child to new child just created
+    parentHierarchy.m_childFirst = child;
+    parentHierarchy.m_childCount ++; // increase child count
+}
+
+void SysHierarchy::cut(ActiveScene& rScene, ActiveEnt ent)
+{
+    auto &rReg = rScene.get_registry();
+    auto &entHier = rReg.get<ACompHierarchy>(ent);
+
+    // TODO: deal with m_depth
+
+    // Unlink siblings by connecting previous and next to each other
+
+    if (rReg.valid(entHier.m_siblingNext))
+    {
+        rReg.get<ACompHierarchy>(entHier.m_siblingNext).m_siblingPrev
+                = entHier.m_siblingPrev;
+    }
+
+    if (rReg.valid(entHier.m_siblingPrev))
+    {
+        rReg.get<ACompHierarchy>(entHier.m_siblingPrev).m_siblingNext
+                = entHier.m_siblingNext;
+    }
+
+    // Unlink parent
+
+    auto &parentHier = rReg.get<ACompHierarchy>(entHier.m_parent);
+    parentHier.m_childCount --;
+
+    if (parentHier.m_childFirst == ent)
+    {
+        parentHier.m_childFirst = entHier.m_siblingNext;
+    }
+
+    entHier.m_level = 0;
+    entHier.m_parent = entHier.m_siblingNext = entHier.m_siblingPrev
+            = entt::null;
+}
+
+void SysHierarchy::sort(ActiveScene& rScene)
+{
+    rScene.get_registry().sort<ACompHierarchy>(
+            [](ACompHierarchy const& lhs, ACompHierarchy const& rhs)
+    {
+        return lhs.m_level < rhs.m_level;
+    }, entt::insertion_sort());
+}
+
+
+void SysHierarchy::update_delete(ActiveScene& rScene)
+{
+    auto &rReg = rScene.get_registry();
+    auto viewDelHier = rReg.view<ACompDelete, ACompHierarchy>();
+
+    // copy entities to delete into a buffer
+    std::vector<ActiveEnt> toDelete;
+    toDelete.reserve(viewDelHier.size_hint());
+    toDelete.assign(std::begin(viewDelHier), std::end(viewDelHier));
```

Comment:
is the call to reserve() actually useful? I suppose i would have expected .assign() to use std::distance() to do a single allocation.